### DISCUSSION TITLE
Official Docker images

### DISCUSSION THREAD

**Discussion by frafra** [ckan/ckan#6492] [Ideas]
Wouldn't it be nice to produce official CKAN Docker images?

**Comment by Zharktas** [comment]
Wouldn't it be nice have funding for maintaining them which the core developers do not have? Currently we are struggling even with have an installation instructions for single OS and keeping up with bugs and new features.

**Reply by frafra** [reply]
I am sorry to hear that. It is a good proposal, but off-topic in my opinion.

As I am not a core developer, but I have been a minor contributor to CKAN (14 PR merged) and some CKAN extensions, I opened a discussion to see if the core developers are OK with my proposal, before writing a PR.

**Reply by Zharktas** [reply]
Contributions are always welcome of course. However "Official docker images" do require maintenance and when there is no allocated time for it they tend to end up unmaintained.

**Reply by frafra** [reply]
I do not agree with that, as the maintenance for Docker images is mostly zero. There is an initial investment in terms of time, as GitHub actions needs to be setup and CI integration (so commits that break Docker images are stopped).

The point here is that everybody which uses CKAN on Docker needs to do that already, in some way or another, including me. I set up a CI system building and publishing CKAN images. I would prefer to work more on that and push that to CKAN because it just makes sense. I would rather have few CKAN bugs assigned for that during the year if something goes wrong than doing that by myself.

**Reply by Zharktas** [reply]
I'm not saying its a ton of work, the initial investment is the problem which you are welcome to contribute :)

However when there is a issue with the official images, someone needs to look into it or the image is just unmaintaned. No one has maintained the Dockerfile in the repository in years and there are almost weekly issues about it. but then again no one has contributed a dockerfile that works for everyone :)

In most cases its just easier to refer to use keitaros images https://github.com/keitaroinc/docker-ckan which they do maintain.

**Reply by frafra** [reply]
Ok, so you are not worried about publishing the result of CKAN Dockerfile, but more about the Dockerfile itself. I made some improvements over time, but I see there are various third-party repositories for that. I wonder why these people felt the need to create something new instead of improving the CKAN Dockerfile.
Many issues could also be prevented by adding some simple end-to-end test on Docker. Bring CKAN up, try to open the dataset page and check HTTP status would probably catch most of the worst issues, and it would be time well spent on my side to avoid future regressions.

If CKAN is not interested in Docker at all, it should deprecate it and suggest external resources, maybe.

**Reply by Zharktas** [reply]
Docker is highly requested feature considering how many issues are posted, its just that the tech team themselves do not run ckan with docker in production, at least i'm not aware of any. And as nothing is preventing running ckan in docker, theres no reason to remove it, it would be nice that someone who knows about running ckan instances would contribute it though. There already were discussion about automating release process here https://github.com/ckan/ckan/discussions/6381 but it'll take some time to get there.

Mostly the stance has been that the ubuntu is the only officially supported environment and if that fails, the issue will be solved. Any other environment and its problems has been delegated to the community as giving support by the tech team is time away from something else. And since the development is mostly done by volunteers, time is always a limited resource.

**Reply by ricardogsilva** [reply]
We are working on a new CKAN-based project that is to be deployed with docker images.

At this stage I'm not volunteering to help out with this effort, but I can suggest our project be looked at as an example of:

- a clean `Dockerfie` - https://github.com/kartoza/ckanext-dalrrd-emc-dcpr/blob/main/Dockerfile
- a `docker-compose.dev.yml` file that is suitable for setting up a local development environment that uses volume binds to mount the repo inside the container - https://github.com/kartoza/ckanext-dalrrd-emc-dcpr/blob/main/docker/docker-compose.dev.yml
- a set of github actions workflows that build the docker image, then run automated tests inside it, then publishes to dockerhub, for each commit and even for PRs coming from forks - https://github.com/kartoza/ckanext-dalrrd-emc-dcpr/tree/main/.github/workflows

Some caveats:

- The project will ship a custom ckan extension, so the `Dockerfile` cannot be considered for vanilla CKAN as is;
- We are using [poetry](https://python-poetry.org/) to install python dependencies;
- Added a custom docker entrypoint script written in Python
- It uses [gunicorn](https://gunicorn.org/) as the application server

I would like to help out with bringing a maintainable docker-based CKAN workflow, but can't really commit to something concrete for now.

**Reply by frafra** [reply]
Thanks @ricardogsilva!

There are good  things that could be merged back indeed. It would also be interesting for me to understand why Kartoza felt the need to do it from scratch instead of improving existing CKAN files, or starting with a fork. Could it be that the lack of manpower on CKAN side makes the mainline process less attractive because of long times to take decision on that and/or accept PR? I am just wondering if something could be improved to avoid having more people in the future reinventing the wheel and ending up with many incompatible versions of the same thing.

Another question arises: who should take care of such effort? I would like to help, but it cannot be a one-man show :)
What about opening an issue for each of the possible improvement that can be mainlined? Like having a dev compose file, importing and customizing the existing GitHub actions that Kartoza made, consider a different entrypoint, and such. Would there be interest from Kartoza to support such effort together with others (including me)? Would that be beneficial for you?

For me, mainlining things has always been crucial for my projects, as well as adding end-to-end tests and some automated build procedures, because they are like an insurance :) I also think it saves time in the long run. I do not know if others have the same ideas or requirements I have.

**Reply by frafra** [reply]
> it would be nice that someone who knows about running ckan instances would contribute it though.

I contributed to it, even if they have been minor contributions :)
What about enabling the GitHub sponsorship button for OKF/CKAN to try to get some funding or at least start doing something in that direction? It wouldn't take much time.

> Mostly the stance has been that the ubuntu is the only officially supported environment and if that fails, the issue will be solved. Any other environment and its problems has been delegated to the community as giving support by the tech team is time away from something else. And since the development is mostly done by volunteers, time is always a limited resource.

If the setup procedure is done properly and shared between the regular setup and the Docker setup (so the changes are almost none), such issue should be almost non-existing. We could just use Ubuntu instead of Debian for the Dockerfile and use the same setup scripts if we want to make it easier to give better support for Docker users.

**Reply by ricardogsilva** [reply]
> It would also be interesting for me to understand why Kartoza felt the need to do it from scratch instead of improving existing CKAN files, or starting with a fork.

We are going with a custom docker-related stack mostly because we realized that the project I mentioned before will require some standard extensions to be installed and we will also need to develop and add our own extension in order to meet business requirements. Also, we're making some opinionated choices too (e.g. basing on debian bullseye, using gunicorn and poetry). So it seemed wiser to go with a custom stack, which allows us more leverage with regard to meeting project deadlines.

The plan is to interact with the CKAN community and eventually propose some modifications to upstream, where it makes sense - with this approach we are able to do it with more time, which usually also means that the eventual changes will be of better quality too :sweat_smile: 


> Would there be interest from Kartoza to support such effort together with others (including me)? Would that be beneficial for you?

Definitely! However, as mentioned above, I'm not really able to commit to something concrete at this stage. For the moment, we'd like to ease into the CKAN community - as this current project progresses we're hoping to reach a position where we'll be able to start providing some help with upstream CKAN duties.

**Reply by GordianDziwis** [reply]
First of all I have to say I am working with and on a lot of open source projects, so I know the struggle and please know that what I am saying comes with the appreciation of our hard work.
But... you are doing containers wrong!

1. Containers should not be another way to install CKAN. Containers should be the way to develop CKAN. This would ensure, that everyone is developing in the same environment with the same dependencies and would reduce the time to get a development environment running dramatically. Which will make it easier for outside contributors. Also all the CI and testing stuff would be simplified.
2. The CKAN team has only to maintain a containerized development deployment, which can be very opinionated, but would be a stable starting point for production deployments.
3. Docker is dead, long live podman. Docker is full of API changes and bugs and docker-compose just plain sucks. Podman pods are a much cleaner and flexible solution to wire a few containers together for development. And a Kubernetes cluster config can be exported from a pod!
4. When the development is container based, the Dockerfile is the install instruction. Documentation would be limited to how to start and configure the containers.
5. I checked out `ckan/ckan-docker` and it looks unmaintainable, there are like 10 different CKAN Dockerfiles, configuration is set during build time etc. 

All of this are reasons, that I will build a minimal CKAN podman development deployment and I will maintain it for the next two years, I would love to coordinate and contribute.

**Reply by frafra** [reply]
1. It should not be mandatory; with robust installation script/procedure, it shouldn't really change much if you install CKAN on an Ubuntu container or on an Ubuntu VM or regular system. I agree that it could be beneficial to propose it as preferred method in the future.
2. +1
3. It shouldn't really matter if you use Podman, Docker or Kubernetes. I use all three, including docker-compose, for different reasons. Using words like "dead" and "sucks" doesn't make your argumentation look more solid.
4. +1
5. I use the one hosted here and try to propose some PR in the meanwhile; little by little :)

**Comment by amercader** [comment]
The issue of "official" Docker images is one that regularly comes up and never gets a satisfactory outcome because a number of reasons:

* "maintenance for Docker images is mostly zero" is simply untrue :) If that were the case the ones included in the repo would work just fine and we wouldn't be having this conversation. Aside from the initial investment, maintenance involves iterative improvement, adapting to user feedback, support new Docker features, testing etc.
* Providing a CKAN image that is generic enough to account for different use cases is hard. Users have many preferences and needs and trying to accommodate them in a single Docker setup is hard. Also there are many ways to skin a Docker cat and everybody has their preferences
* I don't think nobody among the tech team is using the images shipped with the CKAN repo. I can only speak for my case but for our projects we use the [openknowledge/docker-ckan](https://github.com/okfn/docker-ckan) ones because they fit our needs, have a limited scope, we are familiar with them and have very limited time to migrate to something else. I assume that others that use eg [keitaroinc/docker-ckan](https://github.com/keitaroinc/docker-ckan) are in similar situation.
* Well-intentioned contributors that help with the initial setup or punctual fixes don't keep involved with the project going forward (which is 100% understandable and ok) so it falls on the tech team to maintain a setup they don't know or use themselves.

After many years of revisiting the Docker issue I personally think we have two options:

* Option 1: Deprecate and eventually drop the core Docker images, and suggest users a list of curated community projects that we know are somehow well maintained: [keitaroinc/docker-ckan](https://github.com/keitaroinc/docker-ckan), [openknowledge/docker-ckan](https://github.com/okfn/docker-ckan), Kartoza's, others that I'm not familiar with... (I really think here that we should not endorse just one option and give options to users)
* Option 2: Try again to provide an "official" Docker image / setup. In this case images should be moved to their own repo ([ckan/ckan-docker](https://github.com/ckan/ckan-docker)) so they can be maintained by the wider community (with final say from our Docker tzar @kowh-ai). If this was to be successful I think these images should:
   * Be limited in scope, and act as a base that users can extend to their own needs
   * Be opinionated, and provide one way to do things
   * Be automatically tested

Would love to hear people's thoughts on this, from @ckan/core folks and others

**Reply by frafra** [reply]
>     * "maintenance for Docker images is mostly zero" is simply untrue :) If that were the case the ones included in the repo would work just fine and we wouldn't be having this conversation.

There are no Docker images in the repo, there are just Dockerfiles. That is why I need to build them by myself and set up a custom CI procedure for that and such :)

> Aside from the initial investment, maintenance involves iterative improvement, adapting to user feedback, support new Docker features, testing etc.

New Docker features for Dockerfiles? Basically none, syntax is the same. A well-built Dockerfile aligned with the Ubuntu procedure/scripts would be almost zero maintenance. New CKAN-releated features would require some effort indeed.

My idea initially was just "let's build Docker images from Dockerfiles", to avoid having everybody doing the same on their PCs or servers. This is trivial. Then the discussion became more interesting, with possibilities to further improve Docker support and merge with other efforts.

>     * I don't think nobody among the tech team is using the images shipped with the CKAN repo. I can only speak for my case but for our projects we use the [openknowledge/docker-ckan](https://github.com/okfn/docker-ckan) ones because they fit our needs, have a limited scope, we are familiar with them and have very limited time to migrate to something else. I assume that others that use eg [keitaroinc/docker-ckan](https://github.com/keitaroinc/docker-ckan) are in similar situation.

I use the regular one with some custom support for extensions, as I wasn't aware of the others, actually. It could be an idea to add some links or tips for new users.

> After many years of revisiting the Docker issue I personally think we have two options:
> 
>     * Option 1: Deprecate and eventually drop the core Docker images, and suggest users a list of curated community projects [...]
> 
>     * Option 2: Try again to provide an "official" Docker image / setup. [...]

I like option number 2, but even option number 1 would be an improvement in my opinion, in terms of having one less Docker setup around, and it would be easier to understand what the alternatives are.

**Reply by Zharktas** [reply]
Apparently https://hub.docker.com/r/ckan/ckan is built from the https://github.com/ckan/ckan/blob/master/Dockerfile and has been downloaded more than 500k times. But dockerhub only has latest tag updated, so its not really usable in production.
